File: Examples/CSharp/x-IMU3-Examples/Examples/SendCommands.cs
namespace Ximu3Examples
{
    class SendCommands
    {
        public SendCommands()
        {
            // Search for connection
            var devices = Ximu3.PortScanner.ScanFilter(Ximu3.CApi.XIMU3_PortType.XIMU3_PortTypeUsb);

            if (devices.Length == 0)
            {
                Console.WriteLine("No USB connections available");
                return;
            }

            Console.WriteLine("Found " + Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_device_to_string(devices[0])));

            // Open connection
            var connection = new Ximu3.Connection(Ximu3.ConnectionConfig.From(devices[0])!);

            var result = connection.Open();

            if (result != Ximu3.CApi.XIMU3_Result.XIMU3_ResultOk)
            {
                Console.WriteLine("Unable to open " + connection.GetConfig() + ". " + Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_result_to_string(result)) + ".");
                return;
            }

            // Example commands
            string[] commands =
            [
                "{\"device_name\":\"Foobar\"}", // write "Foobar" to device name
                "{\"serial_number\":null}", // read serial number
                "{\"firmware_version\":null}", // read firmware version
                "{\"invalid_key\":null}", // invalid key to demonstrate an error response
            ];

            // Send commands
            if (Helpers.YesOrNo("Use async implementation?"))
            {
                connection.SendCommandsAsync(commands, Callback);

                System.Threading.Thread.Sleep(3000);
            }
            else
            {
                PrintResponses(connection.SendCommands(commands));
            }

            // Close connection
            connection.Close();
        }

        private void Callback(Ximu3.CommandMessage?[] responses)
        {
            PrintResponses(responses);
        }

        private static void PrintResponses(Ximu3.CommandMessage?[] responses)
        {
            foreach (var response in responses)
            {
                if (response == null)
                {
                    Console.WriteLine("No response");
                    continue;
                }

                if (response.Error != null)
                {
                    Console.WriteLine(response.Error);
                    continue;
                }

                Console.WriteLine(response.Key + " : " + response.Value);
            }
        }
    }
}

File: Examples/Rust/src/examples/send_commands.rs
use crate::helpers;
use ximu3::command_message::*;
use ximu3::connection::*;
use ximu3::port_scanner::*;

pub fn run() {
    // Search for connection
    let devices = PortScanner::scan_filter(PortType::Usb);

    std::thread::sleep(std::time::Duration::from_secs(1)); // wait for OS to release port

    let Some(device) = devices.first() else {
        println!("No USB connections available");
        return;
    };

    println!("Found {device}");

    // Open connection
    let connection = Connection::new(&device.connection_config);

    if let Err(error) = connection.open() {
        println!("Unable to open {}. {error}.", connection.get_config());
        return;
    }

    // Example commands
    let commands = vec![
        "{\"device_name\":\"Foobar\"}".into(), // write "Foobar" to device name
        "{\"serial_number\":null}".into(),     // read serial number
        "{\"firmware_version\":null}".into(),  // read firmware version
        "{\"invalid_key\":null}".into(),       // invalid key to demonstrate an error response
    ];

    // Send commands
    if helpers::yes_or_no("Use async implementation?") {
        let closure = Box::new(|responses| {
            print_responses(responses);
        });

        connection.send_commands_async(commands, DEFAULT_RETRIES, DEFAULT_TIMEOUT, closure);

        std::thread::sleep(std::time::Duration::from_secs(3));
    } else {
        print_responses(connection.send_commands(commands, DEFAULT_RETRIES, DEFAULT_TIMEOUT));
    }

    // Close connection
    connection.close();
}

fn print_responses(responses: Vec<Option<CommandMessage>>) {
    for response in responses {
        let Some(response) = response else {
            println!("No response");
            continue;
        };

        if let Some(error) = response.error {
            println!("{error}");
            continue;
        }

        println!("{} : {}", String::from_utf8_lossy(&response.key), String::from_utf8_lossy(&response.value));
    }
}
